src/AspNet.Security.OAuth.Bilibili/BilibiliAuthenticationHandler.cs

```diff
@@ -0,0 +1,218 @@
+    {
+        var parameters = new Dictionary<string, string?>
+        {
+            ["client_id"] = Options.ClientId, // Used instead of "client_id"
+            ["response_type"] = "code",
+            ["gourl"] = redirectUri
+        };
+
+        foreach (var additionalParameter in Options.AdditionalAuthorizationParameters)
+        {
+            parameters.Add(additionalParameter.Key, additionalParameter.Value);
+        }
+
+        parameters["state"] = Options.StateDataFormat.Protect(properties);
+
+        return QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, parameters);
+    }
+
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+    {
+        // See https://open.bilibili.com/doc/4/eaf0e2b5-bde9-b9a0-9be1-019bb455701c#h1-u7B80u4ECB for details.
+        var tokenRequestParameters = new Dictionary<string, string?>()
+        {
+            ["client_id"] = Options.ClientId,
+            ["code"] = context.Code,
+            ["client_secret"] = Options.ClientSecret,
+            ["grant_type"] = "authorization_code",
+        };
+
+        using var tokenRequestContent = new FormUrlEncodedContent(tokenRequestParameters);
+
+        using var response = await Backchannel.PostAsync(Options.TokenEndpoint, tokenRequestContent, Context.RequestAborted);
+
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.AccessTokenError(Logger, response, Context.RequestAborted);
+            return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+        }
+
+        using var stream = await response.Content.ReadAsStreamAsync(Context.RequestAborted);
+        using var document = await JsonDocument.ParseAsync(stream);
+
+        var mainElement = document.RootElement;
+        if (!ValidateReturnCode(mainElement, out var code))
+        {
+            return OAuthTokenResponse.Failed(new Exception($"An error (Code:{code}) occurred while retrieving an access token."));
+        }
+
+        var payload = JsonDocument.Parse(mainElement.GetProperty("data").GetRawText());
+        return OAuthTokenResponse.Success(payload);
+    }
+
+    private static string ComputeHmacSHA256(string key, string data)
+    {
+        using var hmacsha256 = new System.Security.Cryptography.HMACSHA256(System.Text.Encoding.UTF8.GetBytes(key));
+        var hash = hmacsha256.ComputeHash(System.Text.Encoding.UTF8.GetBytes(data));
+        return Convert.ToHexStringLower(hash);
+    }
+
+#pragma warning disable CA5351
+    private static string ComputeMd5(string input)
+    {
+        var inputBytes = Encoding.ASCII.GetBytes(input);
+        var hashBytes = System.Security.Cryptography.MD5.HashData(inputBytes);
+        return Convert.ToHexStringLower(hashBytes).Replace("-", string.Empty, StringComparison.OrdinalIgnoreCase).ToLower(CultureInfo.InvariantCulture);
```

Comment:
`ToLower()` is redundant as you're already using `ToHexStringLower()`, and IIRC, that won't return any dashes either, so the `Replace()` is also redundant.